presto-main/src/main/java/com/facebook/presto/sql/ExpressionUtils.java

```diff
@@ -228,19 +229,21 @@ public static Expression combineDisjunctsWithDefault(Collection<Expression> expr
 
     public static Expression filterDeterministicConjuncts(Expression expression)
     {
-        List<Expression> conjuncts = extractConjuncts(expression).stream()
-                .filter(DeterminismEvaluator::isDeterministic)
-                .collect(toList());
-
-        return combineConjuncts(conjuncts);
+        return filterConjuncts(expression, DeterminismEvaluator::isDeterministic);
     }
 
     public static Expression filterNonDeterministicConjuncts(Expression expression)
     {
-        return combineConjuncts(extractConjuncts(expression)
-                .stream()
-                .filter((conjunct) -> !DeterminismEvaluator.isDeterministic(conjunct))
-                .collect(toImmutableList()));
+        return filterConjuncts(expression, not(DeterminismEvaluator::isDeterministic));
+    }
+
+    public static Expression filterConjuncts(Expression expression, Predicate<Expression> predicate)
```

Comment:
There are no conjuncts in the API (no `List`, just an `expression`). Make it `private`.